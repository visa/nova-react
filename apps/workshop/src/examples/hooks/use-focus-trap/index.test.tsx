/**
 *              Â© 2025 Visa
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 **/
import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { axe } from 'jest-axe';

import metaData from './meta.json';

import { UseFocusTrapExample } from './use-focus-trap-example';

const examples = [{ Component: UseFocusTrapExample, title: metaData['use-focus-trap-example'].title }];

/**
 * Workaround until jest testing environment supports HTMLDialogElement.
 * Issue: https://github.com/jsdom/jsdom/issues/3294
 */
HTMLDialogElement.prototype.show = jest.fn(function mock(this: HTMLDialogElement) {
  this.open = true;
});
HTMLDialogElement.prototype.showModal = jest.fn(function mock(this: HTMLDialogElement) {
  this.open = true;
});
HTMLDialogElement.prototype.close = jest.fn(function mock(this: HTMLDialogElement) {
  this.open = false;
});

describe('useFocusTrap example', () => {
  examples.forEach(({ Component, title }) => {
    it(`${title} should render correctly`, async () => {
      const { container } = render(<Component />);
      const results = await axe(container);
      expect(results).toHaveNoViolations();
      expect(container).toMatchSnapshot();
    });
  });

  describe('dialog', () => {
    it('should open modal', () => {
      const { container } = render(<UseFocusTrapExample />);
      const button = screen.getByText('Open dialog');
      fireEvent.click(button);
      const modal = container.querySelector<HTMLDialogElement>('#use-focus-trap-usage')!;
      expect(modal.hasAttribute('open')).toBe(true);
    });
    it('should close modal', () => {
      const { container } = render(<UseFocusTrapExample />);
      const button = screen.getByText('Open dialog');
      fireEvent.click(button);
      const closeButton = container.querySelector('[aria-label="Close"]')!;
      fireEvent.click(closeButton);
      const modal = container.querySelector<HTMLDialogElement>('#use-focus-trap-usage')!;
      expect(modal.hasAttribute('open')).toBe(false);
    });
    it('should trap focus', async () => {
      render(<UseFocusTrapExample />);
      const button = screen.getByText('Open dialog');
      fireEvent.click(button);
      const primaryButton = screen.getByText('Primary action');
      const user = userEvent.setup();
      await user.tab();
      await user.tab();
      await user.tab();
      await user.tab();
      await user.tab();
      expect(primaryButton).toHaveFocus();
      await user.tab({ shift: true });
      await user.tab({ shift: true });
      await user.tab({ shift: true });
      expect(primaryButton).toHaveFocus();
    });
  });
});
