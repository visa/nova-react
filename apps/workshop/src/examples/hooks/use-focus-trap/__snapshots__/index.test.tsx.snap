// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useFocusTrap example Behavior and usage should render correctly 1`] = `
<div>
  <button
    class="v-button"
  >
    Open dialog
  </button>
  <dialog
    aria-describedby="use-focus-trap-usage-description"
    aria-labelledby="use-focus-trap-usage-title"
    class="v-message v-dialog v-dialog-default"
    id="use-focus-trap-usage"
  >
    <div
      class="v-message-content v-pb-2 v-pr-2"
    >
      <h2
        class="v-dialog-header v-align-items-center v-flex v-gap-6 v-justify-content-start"
        id="use-focus-trap-usage-title"
      >
        Default title
      </h2>
      <p
        id="use-focus-trap-usage-description"
      >
        This is required text that describes the dialog title in more detail.
      </p>
      <div
        class="v-align-items-center v-flex v-flex-wrap v-gap-8 v-justify-content-end v-pt-16"
      >
        <button
          class="v-button"
        >
          Primary action
        </button>
        <button
          class="v-button v-button-secondary"
        >
          Secondary action
        </button>
      </div>
    </div>
    <button
      aria-label="Close"
      class="v-button v-button-small v-button-tertiary v-button-icon v-button-subtle -v-mt-20 -v-ml-12 -v-mr-18"
    >
      <svg
        aria-hidden="true"
        class="v-icon v-icon-visa v-icon-tiny v-icon-close"
        focusable="false"
        height="16"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          class="v-icon-primary"
          d="M8.00001 9.4149L13.9498 15.3646L15.364 13.9504L9.41423 8.00069L15.364 2.05094L13.9498 0.636723L8.00001 6.58647L2.05026 0.636719L0.636047 2.05093L6.5858 8.00069L0.636054 13.9504L2.05027 15.3646L8.00001 9.4149Z"
        />
        <defs>
          <clippath>
            <rect
              height="16"
              transform="translate(0 0.000488281)"
              width="16"
            />
          </clippath>
        </defs>
      </svg>
    </button>
  </dialog>
</div>
`;
